specs/BUNDLE_SPEC.md

```diff
@@ -0,0 +1,271 @@
+# Cosign Bundle Specification
+
+This document aims to describe how `cosign` attaches Sigstore attestation
+[bundles](https://github.com/sigstore/protobuf-specs/blob/main/protos/sigstore_bundle.proto)
+to container images.
+
+The goal is to specify the behavior well enough to promote other implementations
+and enable interoperability. Attestations attached with `cosign` should be
+retrievable in other tools, and vice-versa.
+
+This document focuses on the layout of attestations within an
+[OCI Image Manifest V1.1](https://github.com/opencontainers/image-spec/blob/v1.1.0/manifest.md)
+object.
+
+This document makes no assumptions about the contents of the Sigstore bundle.
+Any attestation which can be represented as a Sigstore bundle (message
+signatures, DSSE-wrapped in-toto statements, etc) can be attached to a container
+image stored in an OCI registry.
+
+Multiple Attestations may be "attached" to one image.
+
+Attestations attached to a container image are generally assumed to refer to
+that image in some way.
+
+## Storage
+
+The approach for storing Sigstore bundles in an OCI registry follows the
+[guidelines for artifact usage](https://github.com/opencontainers/image-spec/blob/main/manifest.md#guidelines-for-artifact-usage)
+in the OCI
+[image spec](https://github.com/opencontainers/image-spec/blob/main/README.md).
+
+### Publishing
+
+First, the bundle itself is stored in its JSON-serialized form as a blob in the
+registry:
+
+```
+POST /v2/foo/blobs/uploads/?digest=cafed00d...
+Content-Type: application/octet-stream
+
+{"mediaType":"application/vnd.dev.sigstore.bundle+json;version=0.2", ...}
+```
+
+In this example “foo” is the name of the repository within the registry to which
+the artifact is being uploaded. The digest included as part of the POST is the
+hex-encoded SHA-256 digest of the raw bytes of the bundle itself.
+
+Once the blob has been created, the next step is to create a manifest that
+associates the bundle blob with the image it describes:
+
+```
+PUT /v2/foo/manifests/sha256:badf00d...
+Content-Type: application/vnd.oci.image.manifest.v1+json
+
+{
+  "mediaType": "application/vnd.oci.image.manifest.v1+json",
+  "schemaVersion": 2,
+  "artifactType": "application/vnd.dev.sigstore.bundle+json;version=0.2",
```

Comment:
You may encounter registries that reject parameters on the media type string. OCI references section 4.2 of the rfc: https://datatracker.ietf.org/doc/html/rfc6838#section-4.2

Reply:
Yeah, I'm not sure what the right answer is here. I've done testing against quite a few registries and, luckily, the only one I've encountered that has issues with this media type is AWS ECR.

I have an [open issue](https://github.com/aws/containers-roadmap/issues/2306) asking the ECR team to loosen the restrictions on accepted media types. And have also [floated the idea](https://github.com/sigstore/protobuf-specs/issues/165) of changing the Sigstore bundle media type.

Reply:
Looks like we're gonna update the Sigstore bundle media type to something which will be compatible with OCI: https://github.com/sigstore/protobuf-specs/pull/279